storage/gcloud/customer_supplied_keys.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Command-line sample app demonstrating customer-supplied encryption keys.
+
+This sample demonstrates uploading an object while supplying an encryption key,
+and retrieving that object's contents using gcloud API.  The sample uses
+the default credential and project.  To review their values, run this command:
+    $ gcloud info
+
+This sample is used on this page:
+    https://cloud.google.com/storage/docs/encryption#customer-supplied
+
+For more information, see the README.md under /storage.
+"""
+
+import argparse
+import base64
+import filecmp
+import os
+import tempfile
+
+from gcloud import storage
+
+# You can (and should) generate your own encryption key. os.urandom(32) is a
+# good way to accomplish this with Python.
+#
+# Although these keys are provided here for simplicity, please remember
+# that it is a bad idea to store your encryption keys in your source code.
+ENCRYPTION_KEY = os.urandom(32)
+
+
+def upload_object(client, bucket_name, filename, object_name, encryption_key):
+    """Uploads an object, specifying a custom encryption key.
+
+    Args:
+        client: gcloud client
+        bucket_name: name of the destination bucket
+        filename: name of file to be uploaded
+        object_name: name of resulting object
+        encryption_key: encryption key to encrypt the object.
+    """
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(object_name)
+    with open(filename, 'rb') as f:
+      blob.upload_from_file(f, encryption_key=encryption_key)
+
+
+def download_object(client, bucket_name, object_name, filename, encryption_key):
+    """Downloads an object protected by a custom encryption key.
+
+    Args:
+        client: gcloud client
+        bucket_name: name of the source bucket
+        object_name: name of the object to be downloaded
+        filename: name of the resulting file
+        encryption_key: the encryption key that the object is encrypted by.
+    """
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(object_name)
+    with open(filename, 'wb') as f:
+        blob.download_to_file(f, encryption_key=encryption_key)
+
```

Comment:
It has to be a future change as gcloud is not currently supporting rewrite afaik.